### PR TITLE
Fixes #29516 - Add Tracer bulk action support

### PR DESCRIPTION
**Pull request by chris1984**
This PR adds the ability to manage traces on multiple host with the bulk action controller.

Video of the functionality:

https://nimbusweb.me/s/share/4143240/6bngqh187rdak5u2ap42

To test you will need 2 CentOS 7 clients either from forklift or manually installing them and remote_execution on the Katello Devel box.

Katello Devel steps to enable remote execution:

In the foreman directory of the katello-devel box create the following file:

`bundler.d/my.local.rb`

Add in the following:

`gem 'foreman_remote_execution'`

Stop the rails server, then run the following commands:
```bash
bundle install
bundle exec rails db:migrate RAILS_ENV=development
bundle exec rails db:seed RAILS_ENV=development
```
Start the rails server back up.

Client steps:

Install subscription-manager

`# yum install subscription-manager`

Register them to your Katello Devel box

Install katello-host-tools and katello-host-tools-tracer from here:

https://theforeman.org/plugins/katello/3.15/installation/clients.html

Run yum update on both of them to get a list of traces for both clients:

`# yum update -y`

Navigate to Hosts - Content Hosts, select both hosts and do Bulk Actions - Manage Traces

Select a few traces and click on Restart Selected

Unfortunately REX is not working with the devel box so the jobs will goto an error state right away.

You can goto Monitor - Jobs and look at the jobs and confirm the selections of services and reboot got handled correctly.

I will file the redmine issues that I have found for the UI for the next triage session.

### CODE REVIEW

**Comment by chris1984** [comment]
Screenshot with REX installed when picking services:

https://nimb.ws/OWy1r6

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    </form>
+
+    <div class="row">
+      <div class="col-sm-12">
+        <div bst-global-notification></div>
+      </div>
+    </div>
+
+    <div data-extend-template="layouts/partials/table.html">
+      <div data-block="search" ng-show=tracesearch></div>
```
Comment (jeremylenz) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:20]: I downloaded and ran [`htmlhint`](https://github.com/htmlhint/HTMLHint) manually and got the following error:
```
L20 |      <div data-block="search" ng-show=tracesearch></div>
                                         ^ The value of attribute [ ng-show ] must be in double quotes. (attr-value-double-quotes)
```

After changing it to `ng-show="tracesearch"` it eliminated the error locally.  Maybe it's not processing that `</div>` because of that?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ * @description
+ *   Provides the functionality for the content host package list and actions.
+ */
+/*jshint camelcase:false*/
+angular.module('Bastion.content-hosts').controller('ContentHostsBulkTracesController',
+    ['$scope', '$state', '$timeout', '$window', '$uibModalInstance', 'HostBulkAction', 'Organization', 'CurrentOrganization', 'Notification', 'Nutupane', 'BastionConfig', 'HostTraces', 'hostIds',
+    function ($scope, $state, $timeout, $window, $uibModalInstance, HostBulkAction, Organization, CurrentOrganization, Notification, Nutupane, BastionConfig, HostTraces, hostIds) {
+        var tracesNutupane;
+
+        tracesNutupane = new Nutupane(HostBulkAction, hostIds, 'hostTraces');
```
Comment (jeremylenz) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:28]: Unless there's some Angular convention I don't know about, it makes more sense to me to declare and assign this on the same line:
```js
var tracesNutupane = new Nutupane(HostBulkAction, hostIds, 'hostTraces');
```

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      <span data-block="list-actions" ng-hide="contentHost.readonly">
+        <div bst-modal="performViaRemoteExecution(true)">
+          <div data-block="modal-header" translate>Restart services on the selected Content Hosts ?</div>
+          <div data-block="modal-body" translate>Are you sure you want to restart the services on the content hosts chosen?</div>
```
Comment (jeremylenz) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:24]: I'm getting a scroll bar on this.  Do you have it too?

modal_confirmation

(if not, it may be related to another PR I had npm installed so don't worry about it)

Comment (jeremylenz) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:24]: "selected content hosts" seems more clear than "content hosts chosen"

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html [hunk 4]
```diff
@@ -0,0 +1,75 @@
+<div data-extend-template="components/views/bst-modal.html">
+  <h4 data-block="modal-header">Content Host Tracer Management</h4>
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:2]: Traces* Management

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    <div class="row">
+      <div class="col-sm-12">
+        <div bst-global-notification></div>
+      </div>
+    </div>
+
+    <div data-extend-template="layouts/partials/table.html">
+      <div data-block="search" ng-show=tracesearch></div>
+      <span data-block="list-actions" ng-hide="contentHost.readonly">
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:21]: i'm guessing 'contentHost.readonly' is undefined because theres no '$scope.contentHost' in the controller (ie can be removed)

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+          <thead>
+            <tr bst-table-head row-select>
+              <th bst-table-column="application" translate>Application</th>
+              <th bst-table-column="app_type" translate>Type</th>
+              <th bst-table-column="helper" translate>Helper</th>
+              <th bst-table-column="affected_hosts" translate>Hostname</th>
+            </tr>
+          </thead>
+          <tbody>
+            <tr bst-table-row ng-repeat="trace in table.rows | filter:table.tracesCompare" row-select="trace">
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:59]: I'm not seeing a `tracesCompare` function in this PR - what does it do for a regular content host?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js (hunk 2 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:28]: Fixed

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html (hunk 1 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:20]: Thanks for looking into this and finding it, everything else was in double quotes so I guess I missed this one when going through it.  Fixed

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html (hunk 3 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:24]: @jeremylenz  Fixed the wording, the scrollbar looks to be part of the CSS from what I was reading online and to remove it, we would have to change some of that. It looks like it is appearing in other modals as well:

https://nimb.ws/bSMB57

I am fine with filing an issue to fix that up and linking it into the pr, but since it is appearing in other modals, I don't think it should block the PR.

File: app/controllers/katello/api/v2/hosts_bulk_actions_controller.rb [hunk 7]
```diff
@@ -225,6 +225,13 @@ def release_version
       respond_for_async :resource => task
     end
 
+    api :POST, "/hosts/bulk/host_traces", N_("Fetch traces on one or more hosts")
```
[the remark this answers is not in the document]
Comment (chris1984) [reply app/controllers/katello/api/v2/hosts_bulk_actions_controller.rb:228]: Fixed

[the remark this answers is not in the document]
Comment (chris1984) [reply app/controllers/katello/api/v2/hosts_bulk_actions_controller.rb:228]: Works for me :)

File: app/controllers/katello/api/v2/hosts_bulk_actions_controller.rb [hunk 8]
```diff
@@ -225,6 +225,13 @@ def release_version
...
+    param_group :bulk_params
+    def host_traces
```
[the remark this answers is not in the document]
Comment (chris1984) [reply app/controllers/katello/api/v2/hosts_bulk_actions_controller.rb:230]: Fixed.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        tracesNutupane.enableSelectAllResults();
+        tracesNutupane.masterOnly = true;
+        $scope.controller = 'host_traces';
+        $scope.table = tracesNutupane.table;
+        $scope.remoteExecutionPresent = BastionConfig.remoteExecutionPresent;
+        $scope.tracesearch = false;
```
[the remark this answers is not in the document]
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:34]: @jturel I am using it here:

https://github.com/Katello/katello/pull/8653/files#diff-517985748979446b6df963ebdab5ade2R20

I had to do this to get rid of the search box, I traced it down and did this work-around.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+                        traceHelpers.push(value.helper);
+                    }
+                }
+            });
+            return traceHelpers;
+        };
+
+        $scope.performViaRemoteExecution = function(customize) {
+            var traceHelpers = $scope.selectedTraceHelpers();
+            var food = _.map($scope.table.getSelected(), 'host_id');
```
[the remark this answers is not in the document]
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:64]: Haha, you caught me ;) fixed.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html (hunk 4 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:2]: Fixed.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html (hunk 3 above)
Comment (jeremylenz) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:24]: @johnpmitsch Could this be related to having the PF4 CSS on our dev boxes?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+                if (value.app_type !== "session") {
+                    if (traceHelpers.indexOf(value.helper) === -1) {
+                        traceHelpers.push(value.helper);
+                    }
+                }
+            });
+            return traceHelpers;
+        };
+
+        $scope.performViaRemoteExecution = function(customize) {
```
[the remark this answers is not in the document]
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:62]: Did your suggestion and removed it from the `function()`

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html (hunk 5 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:21]: Removed

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html (hunk 6 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:59]: In the single content host traces tab there is a search box, I will remove it from here, left over copy and pasta while getting the structure built. Will remove.

**Comment by chris1984** [comment]
@jturel @jeremylenz updated

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    </form>
+
+    <div class="row">
+      <div class="col-sm-12">
+        <div bst-global-notification></div>
+      </div>
+    </div>
+
+    <div data-extend-template="layouts/partials/table.html">
+      <div data-block="search" ng-show="tracesearch"></div>
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:20]: I guess we can just remove this <div> and the $scope.tracesearch since we're not using it. Can be added back should be come along to add that function!

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        $scope.controller = 'host_traces';
+        $scope.table = tracesNutupane.table;
+        $scope.remoteExecutionPresent = BastionConfig.remoteExecutionPresent;
+        $scope.tracesearch = false;
+
+        $scope.traceActionFormValues = {
+            authenticityToken: $window.AUTH_TOKEN.replace(/&quot;/g, '')
+        };
+
+        $scope.selectedTraceHelpers = function() {
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:39]: I'm curious about the behavior here:

If i have two hosts and one of them needs a reboot, will they both end up getting rebooted?

Similarly: if I have two hosts each with different traces, will only the relevant services get restarted on the respective hosts?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+ *   Provides the functionality for the content host package list and actions.
+ */
+/*jshint camelcase:false*/
+angular.module('Bastion.content-hosts').controller('ContentHostsBulkTracesController',
+    ['$scope', '$state', '$timeout', '$window', '$uibModalInstance', 'HostBulkAction', 'Organization', 'CurrentOrganization', 'Notification', 'Nutupane', 'BastionConfig', 'HostTraces', 'hostIds',
+    function ($scope, $state, $timeout, $window, $uibModalInstance, HostBulkAction, Organization, CurrentOrganization, Notification, Nutupane, BastionConfig, HostTraces, hostIds) {
+
+        var tracesNutupane = new Nutupane(HostBulkAction, hostIds, 'traces');
+        tracesNutupane.enableSelectAllResults();
+        tracesNutupane.masterOnly = true;
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:29]: does anything change if we remove this 'masterOnly' line?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+        $scope.controller = 'host_traces';
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:30]: I have a feeling this doesn't do anything - see if things still work without it

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html (hunk 12 above)
Comment (jturel) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:20]: Or do we actually need this empty div in order to hide it? In which case we can just do (I think):

```
<div data-block="search" ng-show="false"></div>
```

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html (hunk 3 above)
Comment (johnpmitsch) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html]: @jeremylenz its possible, but would have to dig in to confirm. You can try pinning back the @theforeman/vendor npm packages to see.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html (hunk 12 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:20]: The `<div data-block="search" ng-show="false"></div>` worked to hide it still, so I will remove the `scope tracesearch` and reference to it

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js (hunk 14 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:29]: Works fine, removing.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js (hunk 15 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:30]: Removed it.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js (hunk 13 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:39]: @jturel good catch it looks like it does not account for that.

Here is a screenshot of me selecting 1 host with a service and 1 host with a reboot:

https://nimb.ws/e8NMA5

And here is what REX is giving us:

https://nimb.ws/bq6OHK

I am guessing that method works fine for a single host which is what I copied from, so I will have to work to adjust it. Would the errata bulk action be a good example to look at or do you suggest a different one?

Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:40]: Fixed in the upcoming revision with your help

**Comment by chris1984** [comment]
[test redmine]

File: config/routes/api/v2.rb [hunk 16]
```diff
@@ -304,6 +304,13 @@ class ActionDispatch::Routing::Mapper
           end
         end
 
+        api_resources :traces, :only => [], :controller => 'host_tracer' do
+          collection do
+            match '/traces' => 'traces#resolve_traces', :via => :get
```
Comment (jturel) [inline config/routes/api/v2.rb:309]: I think this GET route can be removed. The API should be using the PUT route

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+      <div class="col-sm-12">
+        <div bst-global-notification></div>
+      </div>
+    </div>
+
+    <div data-extend-template="layouts/partials/table.html">
+      <div data-block="search" ng-show="false"></div>
+      <span data-block="list-actions">
+        <div bst-modal="performViaRemoteExecution()">
+          <div data-block="modal-header" translate>Restart services on the selected Content Hosts ?</div>
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:16]: What do you think about the text 'Confirm services restart' for the modal header?

File: app/controllers/katello/api/v2/host_tracer_controller.rb [hunk 18]
```diff
@@ -16,6 +16,46 @@ def index
...
+      end
+
+      job_invocations = []
+
+      composers.each do |composer|
+        composer.trigger
+        job_invocations << composer.job_invocation
+      end
+
+      render json: job_invocations
```
Comment (jturel) [inline app/controllers/katello/api/v2/host_tracer_controller.rb:56]: Can you add a test for this method? The assertion might be that we get the right job invocations in the response

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+        $scope.remoteExecutionPresent = BastionConfig.remoteExecutionPresent;
+
+        $scope.performViaRemoteExecution = function() {
+            var traceids = _.map($scope.table.getSelected(), 'id');
+
+            var onSuccess = function () {
+                var message = translate('Successfully initiated restart of services.');
+                Notification.setSuccessMessage(message, {
+                    link: {
+                        children: translate("Click here to view the jobs."),
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:36]: We typically refrain from saying 'click here' since our links are obvious. I vote for rewording this: `View job invocations.`

File: app/controllers/katello/api/v2/host_tracer_controller.rb [hunk 20]
```diff
@@ -16,6 +16,46 @@ def index
       respond_for_index(:collection => collection)
     end
 
+    api :PUT, "/traces/resolve", N_("Resolve Traces")
+    param :trace_ids, Array, :required => true, :desc => N_("Array of Trace IDs")
+    def resolve
+      traces = Katello::HostTracer.where(id: params[:trace_ids])
```
Comment (jturel) [inline app/controllers/katello/api/v2/host_tracer_controller.rb:22]: We need to ensure that that for the given traces, the user has edit permission for the related hosts. We can use our authorization framework for this.

Create a new module `app/models/katello/authorization/host_tracer.rb`
```
module Katello                                                                                                                                                                                
  module Authorization::HostTracer                                                                                                                                                            
    extend ActiveSupport::Concern                                                                                                                                                             
                                                                                                                                                                                              
    include Authorizable                                                                                                                                                                      
                                                                                                                                                                                              
    module ClassMethods                                                                                                                                                                       
      def resolvable                                                                                                                                                                          
        relation = joins_authorized(::Host::Managed, :edit_hosts)                                                                                                                             
        relation = relation.where(organization_id: User.current.organization_ids) unless User.current.admin?                                                                                  
        relation                                                                                                                                                                              
      end                                                                                                                                                                                     
    end                                                                                                                                                                                       
  end                                                                                                                                                                                         
end
```

This code gives us a new scope: `HostTracer.resolvable` which will return an active record relation of traces we can resolve. If we're an admin it returns everything, otherwise it returns the traces we can edit based on our edit_host permissions.

you can include that module in this class and use `Katello::HostTracer.readable.where.....` here

Comment (jturel) [reply app/controllers/katello/api/v2/host_tracer_controller.rb:22]: I haven't tested it but this should be 99% of the way there!

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html (hunk 17 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/views/content-hosts-bulk-traces-modal.html:16]: Fixed

File: config/routes/api/v2.rb (hunk 16 above)
Comment (chris1984) [reply config/routes/api/v2.rb:309]: Fixed

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js (hunk 19 above)
Comment (chris1984) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-traces-modal.controller.js:36]: Fixed

File: app/controllers/katello/api/v2/host_tracer_controller.rb (hunk 20 above)
Comment (chris1984) [reply app/controllers/katello/api/v2/host_tracer_controller.rb:22]: Added and fixed, will submit to jenkins and see if the same permissions tests fails.

**Comment by chris1984** [comment]
[test katello]

**Comment by chris1984** [comment]
@jturel updated :)

File: app/controllers/katello/api/v2/host_tracer_controller.rb [hunk 21]
```diff
@@ -16,6 +16,45 @@ def index
...
+          trace.helper = 'reboot'
+        end
+      end
+
+      traces_by_host_id = traces.group_by(&:host_id)
+      traces_by_helper = traces.group_by(&:helper)
+
+      composers = []
+
+      if traces_by_host_id.size < traces_by_helper.size
```
Comment (jturel) [inline app/controllers/katello/api/v2/host_tracer_controller.rb:35]: None of the tests actually test the code within this `if` statement - just the else. I found it by stepping through the code. I think if we update one of the tests to add an additional trace with a different helper for one of the hosts we already have, that will do the trick

File: test/controllers/api/v2/host_tracer_controller_test.rb [hunk 22]
```diff
@@ -15,10 +15,59 @@ def setup
...
-      results = JSON.parse(get(:index, params: { :host_id => @host.id }).body)
+      @host1.host_traces.create!(:id => 1, :host_id => 1, :helper => 'agile', :app_type => 'foo', :application => 'scrumm')
+
+      results = JSON.parse(get(:index, params: { :host_id => @host1.id }).body)
+
+      assert_response :success
+      assert_includes results['results'].collect { |item| item['id'] }, @host1.host_traces.first.id
+    end
+
+    def test_resolve_group_by_helper
```
Comment (jturel) [inline test/controllers/api/v2/host_tracer_controller_test.rb:26]: These new tests fail on master because the 'JobInvocationComposer' class doesn't exist since we don't have REX there. We can easily work around this by defining the class ourselves near the top of this file - under the `requires`  - if REX is not present, like this:

```
unless Katello.with_remote_execution?                                                                                                                                                         
  class JobInvocationComposer                                                                                                                                                                 
  end                                                                                                                                                                                         
end
```

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/views/content-hosts.html [hunk 23]
```diff
@@ -51,6 +51,10 @@ <h2 translate>Content Hosts</h2>
           <a ng-click="openModuleStreamsModal()" disable-link="table.numSelected === 0" translate>Manage Module Streams</a>
         </li>
 
+        <li role="menuitem" ng-show="permitted('edit_hosts')" ng-class="{disabled: table.numSelected === 0}">
+          <a ng-click="openTracesModal()" disable-link="table.numSelected === 0" translate>Manage Host Traces</a>
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/views/content-hosts.html:55]: This should be the final functional change. In cases where REX is not installed I think this link should be disabled.

In the bulk traces modal controller you added this:

```
$scope.remoteExecutionPresent = BastionConfig.remoteExecutionPresent;
```

I think we actually need that in the controller for this view and to check for it here since there's no point in being able to get into the modal without REX, imo. What do you think?

With this, you can remove all of the 'REX' stuff from the bulk traces controller
